runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkBatchSideInputHandlerFactory.java

```diff
@@ -0,0 +1,131 @@
+import org.apache.beam.model.pipeline.v1.RunnerApi.ExecutableStagePayload.SideInputId;
+import org.apache.beam.runners.core.construction.graph.ExecutableStage;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PCollectionNode;
+import org.apache.beam.runners.core.construction.graph.SideInputReference;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandler;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandlers.MultimapSideInputHandler;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandlers.MultimapSideInputHandlerFactory;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.KV;
+import org.apache.flink.api.common.functions.RuntimeContext;
+
+/**
+ * {@link StateRequestHandler} that uses a Flink {@link RuntimeContext} to access Flink broadcast
+ * variable that represent side inputs.
+ */
+class FlinkBatchSideInputHandlerFactory implements MultimapSideInputHandlerFactory {
+
+  // Map from side input id to global PCollection id.
+  private final Map<SideInputId, PCollectionNode> sideInputToCollection;
+  private final RuntimeContext runtimeContext;
+
+  /**
+   * Creates a new state handler for the given stage. Note that this requires a traversal of the
+   * stage itself, so this should only be called once per stage rather than once per bundle.
+   */
+  static FlinkBatchSideInputHandlerFactory forStage(
+      ExecutableStage stage, RuntimeContext runtimeContext) {
+    ImmutableMap.Builder<SideInputId, PCollectionNode> sideInputBuilder = ImmutableMap.builder();
+    for (SideInputReference sideInput : stage.getSideInputs()) {
+      sideInputBuilder.put(
+          SideInputId.newBuilder()
+              .setTransformId(sideInput.transform().getId())
+              .setLocalName(sideInput.localName())
+              .build(),
+          sideInput.collection());
+    }
+    return new FlinkBatchSideInputHandlerFactory(sideInputBuilder.build(), runtimeContext);
+  }
+
+  private FlinkBatchSideInputHandlerFactory(
+      Map<SideInputId, PCollectionNode> sideInputToCollection, RuntimeContext runtimeContext) {
+    this.sideInputToCollection = sideInputToCollection;
+    this.runtimeContext = runtimeContext;
+  }
+
+  @Override
+  public <K, V, W extends BoundedWindow> MultimapSideInputHandler<K, V, W> forSideInput(
+      String transformId,
+      String sideInputId,
+      Coder<K> keyCoder,
+      Coder<V> valueCoder,
+      Coder<W> windowCoder) {
+    PCollectionNode collectionNode =
+        sideInputToCollection.get(
+            SideInputId.newBuilder().setTransformId(transformId).setLocalName(sideInputId).build());
+    checkArgument(collectionNode != null, "No side input for %s/%s", transformId, sideInputId);
+    List<Object> broadcastVariable = runtimeContext.getBroadcastVariable(collectionNode.getId());
```

Comment:
Can this be a `List<WindowedValue<KV<K, V>>>`?

Reply:
Turns out it can! I assumed this could only be `List<Object>` because that's how it's used elsewhere. Thanks for pointing this out.